src/pal/src/exception/machexception.cpp

```diff
@@ -728,63 +684,55 @@ GetExceptionString(
-   MIG_DESTROY_REQUEST if the error was not handled
+    None
 --*/
 
 static
-kern_return_t
-catch_exception_raise(
-   mach_port_t exception_port,          // [in] Mach Port that is listening to the exception
-   mach_port_t thread,                  // [in] thread the exception happened on
-   mach_port_t task,                    // [in] task the exception happened on
-   exception_type_t exception,          // [in] The type of the exception
-   MACH_EH_TYPE(exception_data_t) code, // [in] A machine dependent array indicating a particular instance of exception
-   mach_msg_type_number_t code_count)   // [in] The size of the buffer (in natural-sized units). 
+void
+HijackFaultingThread(
+    mach_port_t thread,             // [in] thread the exception happened on
+    mach_port_t task,               // [in] task the exception happened on
+    MachMessage &message)           // [in] exception message
 {
-    kern_return_t MachRet;
-#if defined(_X86_)
-    x86_thread_state32_t ThreadState;
-    const thread_state_flavor_t ThreadStateFlavor = x86_THREAD_STATE32;
-#elif defined(_AMD64_)
-    x86_thread_state64_t ThreadState;
-    const thread_state_flavor_t ThreadStateFlavor = x86_THREAD_STATE64;
-#else
-#error Thread state not defined for this architecture
-#endif
-    unsigned int count = sizeof(ThreadState)/sizeof(unsigned);
-    CONTEXT ThreadContext;
-    EXCEPTION_RECORD ExceptionRecord;
-    memset(&ExceptionRecord, 0, sizeof(ExceptionRecord));
-
-    ExceptionRecord.ExceptionCode = exception_from_trap_code(exception, code, code_count, &ExceptionRecord
-#if defined(_AMD64_)
-        , thread
-#endif // _AMD64_
-        );
+    thread_state_flavor_t threadStateFlavor;
+    x86_thread_state_t threadState;
+    EXCEPTION_RECORD exceptionRecord;
+    CONTEXT threadContext;
+    kern_return_t machret;
+    unsigned int count;
+
+    // Fill in the exception record from the exception message
+    ExceptionRecordFromMessage(message, &exceptionRecord);
     
-    // Don't save/restore the debug registers because loading them on OSx causes a 
-    // priviledged instruction fault. The "DE" in CR4 is set.
+    // Get the thread state from the exception message and convert the count of bytes into
+    // the count of ints
+    threadStateFlavor = message.GetThreadStateFlavor();
+    count = message.GetThreadState(threadStateFlavor, (thread_state_t)&threadState, thread) / sizeof(natural_t);
+
 #ifdef _X86_
-    ThreadContext.ContextFlags = CONTEXT_CONTROL | CONTEXT_INTEGER | CONTEXT_SEGMENTS | CONTEXT_FLOATING_POINT | CONTEXT_EXTENDED_REGISTERS
+    threadContext.ContextFlags = CONTEXT_FLOATING_POINT | CONTEXT_EXTENDED_REGISTERS
 #else
-    ThreadContext.ContextFlags = CONTEXT_CONTROL | CONTEXT_INTEGER | CONTEXT_SEGMENTS | CONTEXT_FLOATING_POINT;
+    threadContext.ContextFlags = CONTEXT_FLOATING_POINT;
 #endif
+    // Get just the floating point registers directly from the thread.
+    machret = CONTEXT_GetThreadContextFromPort(thread, &threadContext);
```

Comment:
I assume you are getting the floating point registers this way because they are not in the exception message. It seems it would be worth asserting that the threadStateFlavor retrieved from the message is really x86_THREAD_STATE32 or x86_THREAD_STATE64.